packages/server/src/api/rest/index.ts

```diff
@@ -756,12 +764,111 @@ class RequestHandler extends APIHandlerBase {
             return this.makeUnsupportedModelError(type);
         }
 
-        const { error, attributes, relationships } = this.processRequestBody(type, requestBody, zodSchemas, 'create');
+        const { error, attributes, relationships, operation, matchFields } = this.processRequestBody(
+            type,
+            requestBody,
+            zodSchemas,
+            'create'
+        );
 
         if (error) {
             return error;
         }
 
+        let entity: any;
+
+        if (operation === 'upsert') {
+            entity = await this.runUpsert(typeInfo, type, prisma, modelMeta, attributes, relationships, matchFields);
+        } else if (operation === 'create') {
+            entity = await this.runCreate(typeInfo, type, prisma, attributes, relationships);
+        } else {
+            return this.makeError('invalidPayload');
+        }
+
+        if (entity.status) {
+            return entity;
+        }
```

Comment:
This is not very elegant. I tried throwing the errors from within each of the upsert and create functions, but that caused some weird regressions with a few tests getting "undefined" from the API handler. I also didn't want to inline the upsert path here since it's quite verbose. I may want to refactor this a bit with fresh eyes another day.